File: src/seedpass/core/portable_backup.py
# portable_backup.py
"""Export and import encrypted profile backups."""

from __future__ import annotations

import base64
import json
import re
import logging
import os
import time
import asyncio
from enum import Enum
from pathlib import Path

from .vault import Vault
from .backup import BackupManager
from nostr.client import NostrClient
from utils.key_derivation import (
    derive_index_key,
    EncryptionMode,
)
from .encryption import EncryptionManager
from utils.checksum import json_checksum, canonical_json_dumps
from utils.fingerprint import generate_fingerprint
from .state_manager import StateManager
from .derivation_collisions import find_derivation_collisions
from .errors import ProfileMismatchError

from termcolor import colored

logger = logging.getLogger(__name__)

FORMAT_VERSION = 1
EXPORT_NAME_TEMPLATE = "seedpass_export_{ts}.json"


class PortableMode(Enum):
    """Encryption mode for portable exports."""

    SEED_ONLY = EncryptionMode.SEED_ONLY.value
    NONE = "none"


def _derive_export_key(seed: str) -> bytes:
    """Derive the Fernet key for the export payload."""

    return derive_index_key(seed)


def export_backup(
    vault: Vault,
    backup_manager: BackupManager,
    dest_path: Path | None = None,
    *,
    publish: bool = False,
    parent_seed: str | None = None,
    encrypt: bool = True,
) -> Path:
    """Export the current vault state to a portable file.

    When ``encrypt`` is ``True`` (the default) the payload is encrypted with a
    key derived from the parent seed.  When ``encrypt`` is ``False`` the payload
    is written in plaintext and the wrapper records an ``encryption_mode`` of
    :data:`PortableMode.NONE`.
    """

    if dest_path is None:
        ts = int(time.time())
        dest_dir = vault.fingerprint_dir / "exports"
        dest_dir.mkdir(parents=True, exist_ok=True)
        dest_path = dest_dir / EXPORT_NAME_TEMPLATE.format(ts=ts)

    index_data = vault.load_index()
    canonical = canonical_json_dumps(index_data)

    if encrypt:
        seed = (
            parent_seed
            if parent_seed is not None
            else vault.encryption_manager.decrypt_parent_seed()
        )
        key = _derive_export_key(seed)
        enc_mgr = EncryptionManager(key, vault.fingerprint_dir)
        payload_bytes = enc_mgr.encrypt_data(canonical.encode("utf-8"))
        mode = PortableMode.SEED_ONLY
        cipher = "aes-gcm"
    else:
        payload_bytes = canonical.encode("utf-8")
        mode = PortableMode.NONE
        cipher = "none"

    checksum = json_checksum(index_data)

    wrapper = {
        "format_version": FORMAT_VERSION,
        "created_at": int(time.time()),
        "fingerprint": vault.fingerprint_dir.name,
        "encryption_mode": mode.value,
        "cipher": cipher,
        "checksum": checksum,
        "payload": base64.b64encode(payload_bytes).decode("utf-8"),
    }

    json_bytes = json.dumps(wrapper, indent=2).encode("utf-8")
    dest_path.write_bytes(json_bytes)
    os.chmod(dest_path, 0o600)
    backup_manager._create_additional_backup(dest_path)

    if publish:
        encrypted = vault.encryption_manager.encrypt_data(json_bytes)
        enc_file = dest_path.with_suffix(dest_path.suffix + ".enc")
        enc_file.write_bytes(encrypted)
        os.chmod(enc_file, 0o600)
        try:
            idx = StateManager(vault.fingerprint_dir).state.get("nostr_account_idx", 0)
            client = NostrClient(
                vault.encryption_manager,
                vault.fingerprint_dir.name,
                config_manager=backup_manager.config_manager,
                account_index=idx,
            )
            asyncio.run(client.publish_snapshot(encrypted))
        except Exception:
            logger.error("Failed to publish backup via Nostr", exc_info=True)

    return dest_path


_FINGERPRINT_RE = re.compile(r"^[0-9A-F]{16}$")


def _looks_like_fingerprint(value: str) -> bool:
    """Is this string a SeedPass profile fingerprint?

    ``generate_fingerprint`` produces 16 uppercase hex characters. Anything
    else in a backup's ``fingerprint`` field tells us nothing about which seed
    produced it.
    """
    return bool(_FINGERPRINT_RE.match(value))


def import_backup(
    vault: Vault,
    backup_manager: BackupManager,
    path: Path,
    parent_seed: str | None = None,
    allow_fingerprint_mismatch: bool = False,
) -> None:
    """Import a portable backup file and replace the current index.

    ``allow_fingerprint_mismatch`` permits importing a backup taken from a
    different profile. See the fingerprint check below for why that is not the
    default.
    """

    raw = Path(path).read_bytes()
    if path.suffix.endswith(".enc"):
        raw = vault.encryption_manager.decrypt_data(raw, context=str(path))

    wrapper = json.loads(raw.decode("utf-8"))
    if wrapper.get("format_version") != FORMAT_VERSION:
        raise ValueError("Unsupported backup format")

    # A seed-only backup is bound to its seed implicitly -- the wrong seed
    # cannot decrypt it. A plaintext one carries no such binding: it imports
    # cleanly into any profile, and every derived entry (passwords, SSH, PGP,
    # seeds) is then re-derived from THIS profile's seed, silently producing
    # different secrets than the backup was taken to preserve. Nothing errors;
    # the user simply gets wrong passwords.
    #
    # The comparison is against the fingerprint DERIVED FROM THE SEED being
    # imported under, not against the profile directory's name. In production
    # those are the same string -- a profile directory is named for its
    # fingerprint -- but the question worth asking is "was this backup made
    # from my seed?", and only the seed can answer it. Comparing directory
    # names would refuse a legitimate restore of the same seed into a freshly
    # named profile.
    #
    # Only enforced when the wrapper records something that actually is a
    # fingerprint. A backup carrying some other label gives no evidence about
    # which seed made it, and refusing on no evidence would block restores
    # without making anyone safer.
    source_fingerprint = str(wrapper.get("fingerprint") or "")
    seed_for_check = (
        parent_seed
        if parent_seed is not None
        else vault.encryption_manager.decrypt_parent_seed()
    )
    target_fingerprint = generate_fingerprint(seed_for_check) or ""
    if (
        _looks_like_fingerprint(source_fingerprint)
        and source_fingerprint != target_fingerprint
        and not allow_fingerprint_mismatch
    ):
        raise ProfileMismatchError(
            f"Backup belongs to profile {source_fingerprint}, but the target "
            f"is {target_fingerprint}. Importing it would re-derive every "
            f"secret from the target's seed, silently producing different "
            f"passwords than the ones in the backup. Import into the matching "
            f"profile, or pass allow_fingerprint_mismatch=True if that is "
            f"really what you want."
        )

    mode = wrapper.get("encryption_mode")
    payload = base64.b64decode(wrapper["payload"])

    if mode == PortableMode.SEED_ONLY.value:
        seed = (
            parent_seed
            if parent_seed is not None
            else vault.encryption_manager.decrypt_parent_seed()
        )
        key = _derive_export_key(seed)
        enc_mgr = EncryptionManager(key, vault.fingerprint_dir)
        enc_mgr._legacy_migrate_flag = False
        index_bytes = enc_mgr.decrypt_data(payload, context="backup payload")
    elif mode == PortableMode.NONE.value:
        index_bytes = payload
    else:
        raise ValueError("Unsupported encryption mode")

    index = json.loads(index_bytes.decode("utf-8"))

    checksum = json_checksum(index)
    if checksum != wrapper.get("checksum"):
        raise ValueError("Checksum mismatch")

    # Two different-kind entries at one BIP-85 app-32 index derive the same
    # key. Neither implementation can create that state; it arrives with data,
    # and an import is one of the ways it arrives. Report rather than refuse:
    # the vault is still importable and the user may need it back.
    for collision in find_derivation_collisions(index):
        logger.warning("%s", collision.message)
        print(colored(f"Warning: {collision.message}", "yellow"))

    backup_manager.create_backup()
    vault.save_index(index)

File: js/packages/core/src/vault/portableBackup.ts
/**
 * Portable encrypted profile backups.
 *
 * Parity target: src/seedpass/core/portable_backup.py, format_version 1.
 * The wrapper is plain JSON; the payload is canonical JSON of the index,
 * either plaintext ("none") or V3 AES-GCM encrypted with the index key of
 * the parent seed ("seed-only"). The checksum is SHA-256 over the canonical
 * JSON of the index and is verified on import.
 */

import { base64 } from "@scure/base";
import { z } from "zod";
import { canonicalJson, canonicalHash } from "../sync/canonical.js";
import { deriveIndexKeyBytes } from "./indexKey.js";
import { decryptPayload } from "./payload.js";
import { encryptV3 } from "./aead.js";
import { utf8 } from "../util/bytes.js";

export const PORTABLE_FORMAT_VERSION = 1;

export const portableBackupSchema = z
  .object({
    format_version: z.number().int(),
    created_at: z.number().int(),
    fingerprint: z.string(),
    encryption_mode: z.enum(["seed-only", "none"]),
    cipher: z.string(),
    checksum: z.string().regex(/^[0-9a-f]{64}$/),
    payload: z.string(),
  })
  .loose();

export type PortableBackup = z.infer<typeof portableBackupSchema>;

export class BackupImportError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "BackupImportError";
  }
}

/**
 * Parse and validate a backup wrapper without decrypting or verifying it.
 *
 * Callers need the wrapper's metadata — `encryption_mode` to know whether a
 * seed is required, `fingerprint` to check the backup belongs to the profile
 * being imported into — *before* they can decide how to import. Reading those
 * by pattern-matching the raw file text works until a payload happens to
 * contain the same bytes; this parses the envelope instead.
 *
 * Structural validation only. A wrapper that parses here can still fail
 * importBackup on checksum, format version, or decryption.
 */
export function parseBackupWrapper(wrapperJson: string | Uint8Array): PortableBackup {
  const text =
    typeof wrapperJson === "string" ? wrapperJson : new TextDecoder().decode(wrapperJson);
  try {
    return portableBackupSchema.parse(JSON.parse(text));
  } catch (e) {
    throw new BackupImportError(`invalid backup wrapper: ${String(e)}`);
  }
}

/** Import a portable backup wrapper, returning the verified index payload. */
export async function importBackup(
  wrapperJson: string | Uint8Array,
  options: { mnemonic?: string } = {},
): Promise<Record<string, unknown>> {
  const wrapper = parseBackupWrapper(wrapperJson);
  if (wrapper.format_version !== PORTABLE_FORMAT_VERSION) {
    throw new BackupImportError("Unsupported backup format");
  }

  const payload = base64.decode(wrapper.payload);
  let indexBytes: Uint8Array;
  if (wrapper.encryption_mode === "seed-only") {
    if (!options.mnemonic) {
      throw new BackupImportError("parent seed required for seed-only backups");
    }
    const key = deriveIndexKeyBytes(options.mnemonic);
    try {
      indexBytes = await decryptPayload(key, payload);
    } catch {
      throw new BackupImportError("failed to decrypt backup payload");
    }
  } else {
    indexBytes = payload;
  }

  let index: Record<string, unknown>;
  try {
    index = JSON.parse(new TextDecoder().decode(indexBytes)) as Record<string, unknown>;
  } catch {
    throw new BackupImportError("backup payload is not valid JSON");
  }

  if (canonicalHash(index) !== wrapper.checksum) {
    throw new BackupImportError("Checksum mismatch");
  }
  return index;
}

/** Export an index as a portable backup wrapper (format_version 1). */
export async function exportBackup(
  index: Record<string, unknown>,
  options: {
    mnemonic: string;
    fingerprint: string;
    encrypt?: boolean;
    createdAt?: number;
    /** Fixture/testing only — production draws a random nonce. */
    nonce?: Uint8Array;
  },
): Promise<PortableBackup> {
  const canonical = canonicalJson(index);
  const checksum = canonicalHash(index);
  const encrypt = options.encrypt ?? true;

  let payloadBytes: Uint8Array;
  let mode: "seed-only" | "none";
  let cipher: string;
  if (encrypt) {
    const key = deriveIndexKeyBytes(options.mnemonic);
    payloadBytes = await encryptV3(key, utf8(canonical), options.nonce);
    mode = "seed-only";
    cipher = "aes-gcm";
  } else {
    payloadBytes = utf8(canonical);
    mode = "none";
    cipher = "none";
  }

  return {
    format_version: PORTABLE_FORMAT_VERSION,
    created_at: options.createdAt ?? Math.floor(Date.now() / 1000),
    fingerprint: options.fingerprint,
    encryption_mode: mode,
    cipher,
    checksum,
    payload: base64.encode(payloadBytes),
  };
}
